server/src/main/java/org/cloudfoundry/identity/uaa/scim/jdbc/JdbcScimUserProvisioning.java

```diff
@@ -172,6 +184,77 @@ public List<ScimUser> retrieveByUsernameAndZone(String username, String zoneId)
         return jdbcTemplate.query(USER_BY_USERNAME_AND_ZONE_QUERY , mapper, username, zoneId);
     }
 
+    @Override
+    public List<ScimUser> retrieveByScimFilterOnlyActive(
+            final String filter,
+            final String sortBy,
+            final boolean ascending,
+            final String zoneId
+    ) {
+        /* We cannot reuse the query converter from the superclass here since the later query operates on both the
+         * "users" and the "identity_provider" table and they both have a column named "id". Since the SCIM filter might
+         * contain clauses on the "id" field, we must ensure that the "id" of the "users" table is used, which is done
+         * by attaching an AttributeNameMapper. */
+        final SimpleSearchQueryConverter queryConverter = new SimpleSearchQueryConverter();
+
+        // ensure that the generated query handles the case-insensitivity of the underlying DB correctly
+        queryConverter.setDbCaseInsensitive(useCaseInsensitiveQueries);
+
+        validateOrderBy(queryConverter.map(sortBy));
+
+        /* since the two tables used in the query ('users' and 'identity_provider') have columns with identical names,
+         * we must ensure that the columns of 'users' are used in the WHERE clause generated for the SCIM filter */
+        final AttributeNameMapper attributeNameMapper = new AttributeNameMapper() {
+            @Override
+            public String mapToInternal(final String attr) {
+                // in the later query, 'users' will have the alias 'u'
+                return "u." + attr;
+            }
+
+            @Override
+            public String[] mapToInternal(final String[] attr) {
+                return Stream.of(attr).map(this::mapToInternal).toArray(String[]::new);
+            }
+
+            @Override
+            public String mapFromInternal(final String attr) {
+                return attr.substring(2);
+            }
+
+            @Override
+            public String[] mapFromInternal(final String[] attr) {
+                return Stream.of(attr).map(this::mapFromInternal).toArray(String[]::new);
+            }
+        };
+        queryConverter.setAttributeNameMapper(attributeNameMapper);
+
+        // build WHERE clause
+        final ProcessedFilter where = queryConverter.convert(filter, sortBy, ascending, zoneId);
+        final String whereClauseScimFilter = where.getSql();
```

Comment:
I was surprised to see that the deep nesting has been there a long time, and wasn't the cause of the performance degradation. But still it would be much easier to read the SQL without all the nesting.